File: tests/mock_mcp_server.nim
import std/[json, os, strutils]

proc main() =
  for line in stdin.lines:
    let trimmed = line.strip()
    if trimmed.len == 0:
      continue

    let req = parseJson(trimmed)
    let msgId = req{"id"}
    if msgId.isNil or msgId.kind == JNull:
      continue

    let id = msgId
    let meth = req{"method"}.getStr("")

    var response: JsonNode

    case meth
    of "initialize":
      response = %*{"jsonrpc": "2.0", "id": id, "result": {"protocolVersion": "2024-11-05", "serverInfo": {"name": "mock-mcp", "version": "1.0.0"}, "capabilities": {"tools": {}}}}
    of "tools/list":
      response = %*{"jsonrpc": "2.0", "id": id, "result": {"tools": [{"name": "echo", "description": "Echo back a message"}, {"name": "add", "description": "Add two numbers"}, {"name": "greet", "description": "Greet someone"}]}}
    of "tools/call":
      let params = req{"params"}
      let name = params{"name"}.getStr("")
      let toolArgs = params{"arguments"}
      case name
      of "echo":
        response = %*{"jsonrpc": "2.0", "id": id, "result": {"content": [{"type": "text", "text": toolArgs{"message"}.getStr("")}]}}
      of "add":
        let a = toolArgs{"a"}.getInt(0)
        let b = toolArgs{"b"}.getInt(0)
        response = %*{"jsonrpc": "2.0", "id": id, "result": {"content": [{"type": "text", "text": $(a + b)}]}}
      of "greet":
        let nameParam = toolArgs{"name"}.getStr("")
        response = %*{"jsonrpc": "2.0", "id": id, "result": {"content": [{"type": "text", "text": "Hello, " & nameParam & "!"}]}}
      of "image_tool":
        response = %*{"jsonrpc": "2.0", "id": id, "result": {"content": [{"type": "image", "data": "iVBORw0KGgo", "mimeType": "image/png"}, {"type": "text", "text": "image generated"}]}}
      of "error_tool":
        response = %*{"jsonrpc": "2.0", "id": id, "result": {"content": [], "isError": true}}
      of "empty_tool":
        response = %*{"jsonrpc": "2.0", "id": id, "result": {"content": [], "isError": false}}
      else:
        response = %*{"jsonrpc": "2.0", "id": id, "error": {"code": -32601, "message": "Unknown tool: " & name}}
    of "ping":
      response = %*{"jsonrpc": "2.0", "id": id, "result": {}}
    else:
      response = %*{"jsonrpc": "2.0", "id": id, "error": {"code": -32601, "message": "Unknown method: " & meth}}

    echo $response

when isMainModule:
  main()

File: core/src/bin/mock_mcp_server.rs
use serde_json::{json, Value};
use std::io::{self, BufRead, Write};

fn main() {
    let stdin = io::stdin();
    for line in stdin.lock().lines() {
        let trimmed = match line {
            Ok(l) => l.trim().to_string(),
            Err(_) => break,
        };
        if trimmed.is_empty() {
            continue;
        }

        let req: Value = match serde_json::from_str(&trimmed) {
            Ok(v) => v,
            Err(_) => continue,
        };

        let msg_id = &req["id"];
        if msg_id.is_null() {
            continue;
        }

        let id = msg_id.clone();
        let method = req["method"].as_str().unwrap_or("").to_string();

        let response = match method.as_str() {
            "initialize" => json!({
                "jsonrpc": "2.0", "id": id,
                "result": {
                    "protocolVersion": "2024-11-05",
                    "serverInfo": {"name": "mock-mcp", "version": "1.0.0"},
                    "capabilities": {"tools": {}}
                }
            }),
            "tools/list" => json!({
                "jsonrpc": "2.0", "id": id,
                "result": {
                    "tools": [
                        {"name": "echo", "description": "Echo back a message"},
                        {"name": "add", "description": "Add two numbers"},
                        {"name": "greet", "description": "Greet someone"}
                    ]
                }
            }),
            "tools/call" => {
                let params = &req["params"];
                let name = params["name"].as_str().unwrap_or("");
                let tool_args = &params["arguments"];
                match name {
                    "echo" => json!({
                        "jsonrpc": "2.0", "id": id,
                        "result": {
                            "content": [{"type": "text", "text": tool_args["message"].as_str().unwrap_or("")}]
                        }
                    }),
                    "add" => {
                        let a = tool_args["a"].as_i64().unwrap_or(0);
                        let b = tool_args["b"].as_i64().unwrap_or(0);
                        json!({
                            "jsonrpc": "2.0", "id": id,
                            "result": {
                                "content": [{"type": "text", "text": (a + b).to_string()}]
                            }
                        })
                    }
                    "greet" => {
                        let name_param = tool_args["name"].as_str().unwrap_or("");
                        json!({
                            "jsonrpc": "2.0", "id": id,
                            "result": {
                                "content": [{"type": "text", "text": format!("Hello, {}!", name_param)}]
                            }
                        })
                    }
                    "image_tool" => json!({
                        "jsonrpc": "2.0", "id": id,
                        "result": {
                            "content": [
                                {"type": "image", "data": "iVBORw0KGgo", "mimeType": "image/png"},
                                {"type": "text", "text": "image generated"}
                            ]
                        }
                    }),
                    "error_tool" => json!({
                        "jsonrpc": "2.0", "id": id,
                        "result": {"content": [], "isError": true}
                    }),
                    "empty_tool" => json!({
                        "jsonrpc": "2.0", "id": id,
                        "result": {"content": [], "isError": false}
                    }),
                    _ => json!({
                        "jsonrpc": "2.0", "id": id,
                        "error": {"code": -32601, "message": format!("Unknown tool: {}", name)}
                    }),
                }
            }
            "ping" => json!({
                "jsonrpc": "2.0", "id": id,
                "result": {}
            }),
            _ => json!({
                "jsonrpc": "2.0", "id": id,
                "error": {"code": -32601, "message": format!("Unknown method: {}", method)}
            }),
        };

        let output = response.to_string();
        let _ = writeln!(io::stdout(), "{}", output);
        let _ = io::stdout().flush();
    }
}